src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -1465,26 +2499,173 @@ public static TSource ElementAtOrDefault<TSource>(this IEnumerable<TSource> sour
 
         public static IEnumerable<int> Range(int start, int count)
         {
-            long max = ((long)start) + count - 1;
-            if (count < 0 || max > Int32.MaxValue) throw Error.ArgumentOutOfRange("count");
-            return RangeIterator(start, count);
+            return new RangeIterator(start, count);
         }
 
-        private static IEnumerable<int> RangeIterator(int start, int count)
+        private sealed class RangeIterator : ReadOnlyList<int>, IListIterator<int>
```

Comment:
Range and Repeat are perhaps the cleanest here, with no costs in any case over the previous.